File: demo/src/main/java/com/training/demo/service/PaymentService.java
package com.training.demo.service;

import com.training.demo.dto.request.Payment.ConfirmPaymentRequest;
import com.training.demo.dto.request.Payment.CreatePaymentRequest;
import com.training.demo.dto.response.Payment.PaymentResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import java.util.List;

public interface PaymentService {

    /**
     * Lấy tất cả payments (ADMIN)
     */
    Page<PaymentResponse> getAllPayments(Pageable pageable);

    /**
     * Tạo payment mới
     */
    PaymentResponse createPayment(Long userId, CreatePaymentRequest request);

    /**
     * Xác nhận thanh toán (ADMIN hoặc auto từ gateway)
     */
    PaymentResponse confirmPayment(Long paymentId, ConfirmPaymentRequest request);

    /**
     * Lấy chi tiết payment
     */
    PaymentResponse getPaymentById(Long paymentId);

    /**
     * Lấy lịch sử thanh toán của order
     */
    List<PaymentResponse> getOrderPayments(Long orderId);

    /**
     * Lấy payment theo transaction ID
     */
    PaymentResponse getPaymentByTransactionId(String transactionId);

    /**
     * Hủy payment
     */
    PaymentResponse cancelPayment(Long paymentId);
}

File: src/app/core/services/payments/payment.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';
import { BaseResponse } from '../../models/response/base-response';
import { CreatePaymentRequest } from '../../models/request/Payment/CreatePaymentRequest';
import { ConfirmPaymentRequest } from '../../models/request/Payment/ConfirmPaymentRequest';
import { PaymentResponse } from '../../models/response/Payment/PaymentResponse';

@Injectable({
  providedIn: 'root'
})
export class PaymentService {

  private apiUrl = `${environment.apiUrl}/payments`;

  constructor(private http: HttpClient) { }

  /**
   * Tạo payment cho order
   */
  createPayment(request: CreatePaymentRequest): Observable<BaseResponse<PaymentResponse>> {
    return this.http.post<BaseResponse<PaymentResponse>>(this.apiUrl, request);
  }

  /**
   * Xác nhận thanh toán (ADMIN)
   */
  confirmPayment(paymentId: number, request: ConfirmPaymentRequest): Observable<BaseResponse<PaymentResponse>> {
    return this.http.patch<BaseResponse<PaymentResponse>>(`${this.apiUrl}/${paymentId}/confirm`, request);
  }

  /**
   * Lấy chi tiết payment
   */
  getPaymentById(paymentId: number): Observable<BaseResponse<PaymentResponse>> {
    return this.http.get<BaseResponse<PaymentResponse>>(`${this.apiUrl}/${paymentId}`);
  }

  /**
   * Lấy lịch sử thanh toán của order
   */
  getOrderPayments(orderId: number): Observable<BaseResponse<PaymentResponse[]>> {
    return this.http.get<BaseResponse<PaymentResponse[]>>(`${this.apiUrl}/order/${orderId}`);
  }

  /**
   * Lấy payment theo transaction ID
   */
  getPaymentByTransaction(transactionId: string): Observable<BaseResponse<PaymentResponse>> {
    return this.http.get<BaseResponse<PaymentResponse>>(`${this.apiUrl}/transaction/${transactionId}`);
  }

  /**
   * Hủy payment
   */
  cancelPayment(paymentId: number): Observable<BaseResponse<PaymentResponse>> {
    return this.http.patch<BaseResponse<PaymentResponse>>(`${this.apiUrl}/${paymentId}/cancel`, {});
  }

  // ========== ADMIN ENDPOINTS ==========

  /**
   * Lấy tất cả payments với phân trang (ADMIN)
   */
  getAllPayments(page: number = 0, size: number = 10, sortBy: string = 'createdAt', sortDirection: string = 'DESC'): Observable<BaseResponse<any>> {
    return this.http.get<BaseResponse<any>>(`${this.apiUrl}/admin/all`, {
      params: { 
        page: page.toString(), 
        size: size.toString(),
        sortBy: sortBy,
        sortDirection: sortDirection
      }
    });
  }

  /**
   * Đếm tổng số thanh toán (ADMIN)
   */
  countAllPayments(): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/admin/count`);
  }

  /**
   * Tính tổng doanh thu (ADMIN)
   */
  getTotalRevenue(): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/admin/total-revenue`);
  }
}
